openassessment/fileupload/api.py

```diff
@@ -29,3 +37,299 @@ def remove_file(key):
     Remove file from the storage
     """
     return backends.get_backend().remove_file(key)
+
+
+def get_student_file_key(student_item_dict, index=0):
+    """
+    Args:
+        student_item_dict: A dictionary containing keys ('student_id', 'course_id', 'item_id').
+        index (int, optional): The index of a file.
+    """
+    key_template = '{student_id}/{course_id}/{item_id}'
+    index = int(index)
+    if index > 0:
+        key_template += '/{index}'
+    return key_template.format(index=index, **student_item_dict)
+
+
+def _safe_load_json_list(field):
+    """
+    Tries to load JSON-ified string,
+    returns an empty list if we try to load some non-JSON-encoded string.
+    """
+    try:
+        return json.loads(field)
+    except ValueError:
+        return []
+
+
+class FileUpload(object):
+    """
+    A layer of abstraction over the various components of file
+    data stored as ORA XBlock user-scoped fields.
+    """
+    def __init__(self, name=None, description=None, size=None, index=None, descriptionless=False, **student_item_dict):
+        self.name = name
+        self.description = description
+        self.size = size
+        self.index = index
+        self.student_id = student_item_dict.get('student_id')
+        self.course_id = student_item_dict.get('course_id')
+        self.item_id = student_item_dict.get('item_id')
+        self.descriptionless = descriptionless
+
+    @property
+    def exists(self):
+        return (self.description is not None) or self.descriptionless
+
+    @property
+    def download_url(self):
+        if self.exists:
+            try:
+                return get_download_url(self.key)
+            except FileUploadError:
+                logger.error('Error retrieving download URL for key {}.'.format(self.key))
+                return ''
+
+    @property
+    def key(self):
+        """
+        Simple utility method to generate a common file upload key based on
+        the student item.
+
+        Returns:
+            A string representation of the key.
+        """
+        student_item_dict = {
+            'student_id': self.student_id,
+            'course_id': self.course_id,
+            'item_id': self.item_id,
+        }
+        return get_student_file_key(student_item_dict, index=self.index)
+
+    def url_descriptor_tuple(self):
+        """
+        Used in the response template context to provide a file URL, description, and name
+        to render in the client.
+        """
+        return (self.download_url, self.description, self.name)
+
+    def _to_dict(self):
+        """
+        Returns:
+            A dictionary representation of the FileUpload.
+        """
+        attrs = ('description', 'name', 'size', 'course_id', 'student_id', 'item_id', 'descriptionless')
+        return {
+            key: getattr(self, key, None) for key in attrs
+        }
+
+    def __eq__(self, other):
```

Comment:
I like the thinking ahead. Do we foresee a need to compare two file uploads

Reply:
This only exists for the sake of unit tests at the moment.